src/main/java/org/springframework/retry/support/MetricsRetryListener.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.retry.support;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.Timer;
+
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.RetryContext;
+import org.springframework.retry.RetryListener;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link RetryListener} implementation for Micrometer {@link Timer}s around retry
+ * operations.
+ * <p>
+ * The {@link Timer#start} is called from the {@link #open(RetryContext, RetryCallback)}
+ * and stopped in the {@link #close(RetryContext, RetryCallback, Throwable)}. This
+ * {@link Timer.Sample} is associated with the provided {@link RetryCallback} to make this
+ * {@link MetricsRetryListener} instance reusable for many retry operation.
+ * <p>
+ * The registered {@value #TIMER_NAME} {@link Timer} has these tags by default:
+ * <ul>
+ * <li>{@code name} - {@link RetryCallback#getLabel()}</li>
+ * <li>{@code retry.count} - the number of attempts - 1; essentially the successful first
+ * call means no counts</li>
+ * <li>{@code exception} - the thrown back to the caller (after all the retry attempts)
+ * exception class name</li>
+ * </ul>
+ * <p>
+ * The {@link #setCustomTags(Map)} and {@link #setCustomTagsProvider(Function)} can be
+ * used to further customize tags on the timers.
+ *
+ * @author Artem Bilan
+ * @since 2.0.8
+ */
+public class MetricsRetryListener implements RetryListener {
+
+	public static final String TIMER_NAME = "spring.retry";
+
+	private final MeterRegistry meterRegistry;
+
+	private final Map<RetryCallback<?, ?>, Timer.Sample> retryContextToSample = new HashMap<>();
+
+	private final Map<String, String> customTags = new HashMap<>();
```

Comment:
Yeah... Thought about an XML config support where use builder pattern is hard enough.
Plus `Map` API is familiar for users.
But yes, `Tags` is good enough to expose instead for consistency.

Reply:
You can also keep `Tags` as a `private` field and the ctor/setter parameter can be a `Map` (in addition `Tags` can be there too as an overload).

Reply:
Right. But since `Map` and `Tags` are same matter from end-user perspective, I'd prefer to not give them a paradox of choice.
If there no more feedback, I'll go ahead and incorporate what we have agreed upon so far.
Thanks